### DISCUSSION TITLE
3.7.3 to 4.1.1 Upgrade Issues. (RHEL 9.4)

### DISCUSSION THREAD

**Discussion by bytelock-id** [netbox-community/netbox#17600] [Help Wanted!] [answered]
I've got an issue right out of the gate upgrading to 4.1.1 from 3.7.3 looks like an issue with some missing files and psycopg this is a production RHEL 9.4 machine with the minimums installed required.  Just reading some of the others out there having issues with upgrades I'm not sure the direction of the issue since most folks are running ubuntu and not RHEL 9.4 from what I could see.

---------------------
`
[root@netbox netbox]# sudo PYTHON=/usr/bin/python3.11 ./upgrade.sh
You are installing (or upgrading to) NetBox version 4.1.1
Using Python 3.11.7
Creating a new virtual environment at /opt/netbox-4.1.1/venv...
Updating pip (pip install --upgrade pip)...
Requirement already satisfied: pip in ./venv/lib64/python3.11/site-packages (22.3.1)
Collecting pip
  Downloading pip-24.2-py3-none-any.whl (1.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 8.4 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 22.3.1
    Uninstalling pip-22.3.1:
      Successfully uninstalled pip-22.3.1
Successfully installed pip-24.2
pip 24.2 from /opt/netbox-4.1.1/venv/lib64/python3.11/site-packages/pip (python 3.11)
Installing Python system packages (pip install wheel)...
Collecting wheel
  Downloading wheel-0.44.0-py3-none-any.whl.metadata (2.3 kB)
Downloading wheel-0.44.0-py3-none-any.whl (67 kB)
Installing collected packages: wheel
Successfully installed wheel-0.44.0
Installing core dependencies (pip install -r requirements.txt)...
Collecting Django==5.0.9 (from -r requirements.txt (line 1))
  Downloading Django-5.0.9-py3-none-any.whl.metadata (4.1 kB)
Collecting django-cors-headers==4.4.0 (from -r requirements.txt (line 2))
  Downloading django_cors_headers-4.4.0-py3-none-any.whl.metadata (16 kB)
Collecting django-debug-toolbar==4.4.6 (from -r requirements.txt (line 3))
  Downloading django_debug_toolbar-4.4.6-py3-none-any.whl.metadata (3.9 kB)
Collecting django-filter==24.3 (from -r requirements.txt (line 4))
  Downloading django_filter-24.3-py3-none-any.whl.metadata (5.2 kB)
Collecting django-htmx==1.19.0 (from -r requirements.txt (line 5))
  Downloading django_htmx-1.19.0-py3-none-any.whl.metadata (2.7 kB)
Collecting django-graphiql-debug-toolbar==0.2.0 (from -r requirements.txt (line 6))
  Using cached django_graphiql_debug_toolbar-0.2.0-py3-none-any.whl.metadata (3.5 kB)
Collecting django-mptt==0.16.0 (from -r requirements.txt (line 7))
  Downloading django_mptt-0.16.0-py3-none-any.whl.metadata (5.2 kB)
Collecting django-pglocks==1.0.4 (from -r requirements.txt (line 8))
  Using cached django-pglocks-1.0.4.tar.gz (4.1 kB)
  Preparing metadata (setup.py) ... done
Collecting django-prometheus==2.3.1 (from -r requirements.txt (line 9))
  Using cached django_prometheus-2.3.1-py2.py3-none-any.whl.metadata (9.4 kB)
Collecting django-redis==5.4.0 (from -r requirements.txt (line 10))
  Using cached django_redis-5.4.0-py3-none-any.whl.metadata (32 kB)
Collecting django-rich==1.11.0 (from -r requirements.txt (line 11))
  Downloading django_rich-1.11.0-py3-none-any.whl.metadata (7.8 kB)
Collecting django-rq==2.10.2 (from -r requirements.txt (line 12))
  Downloading django_rq-2.10.2-py2.py3-none-any.whl.metadata (18 kB)
Collecting django-taggit==6.0.0 (from -r requirements.txt (line 13))
  Downloading django_taggit-6.0.0-py3-none-any.whl.metadata (3.5 kB)
Collecting django-tables2==2.7.0 (from -r requirements.txt (line 14))
  Using cached django_tables2-2.7.0-py2.py3-none-any.whl.metadata (3.9 kB)
Collecting django-timezone-field==7.0 (from -r requirements.txt (line 15))
  Downloading django_timezone_field-7.0-py3-none-any.whl.metadata (12 kB)
Collecting djangorestframework==3.15.2 (from -r requirements.txt (line 16))
  Downloading djangorestframework-3.15.2-py3-none-any.whl.metadata (10 kB)
Collecting drf-spectacular==0.27.2 (from -r requirements.txt (line 17))
  Downloading drf_spectacular-0.27.2-py3-none-any.whl.metadata (14 kB)
Collecting drf-spectacular-sidecar==2024.7.1 (from -r requirements.txt (line 18))
  Downloading drf_spectacular_sidecar-2024.7.1-py3-none-any.whl.metadata (3.4 kB)
Collecting feedparser==6.0.11 (from -r requirements.txt (line 19))
  Using cached feedparser-6.0.11-py3-none-any.whl.metadata (2.4 kB)
Collecting gunicorn==23.0.0 (from -r requirements.txt (line 20))
  Downloading gunicorn-23.0.0-py3-none-any.whl.metadata (4.4 kB)
Collecting Jinja2==3.1.4 (from -r requirements.txt (line 21))
  Downloading jinja2-3.1.4-py3-none-any.whl.metadata (2.6 kB)
Collecting Markdown==3.7 (from -r requirements.txt (line 22))
  Downloading Markdown-3.7-py3-none-any.whl.metadata (7.0 kB)
Collecting mkdocs-material==9.5.34 (from -r requirements.txt (line 23))
  Downloading mkdocs_material-9.5.34-py3-none-any.whl.metadata (17 kB)
Collecting mkdocstrings==0.26.1 (from mkdocstrings[python-legacy]==0.26.1->-r requirements.txt (line 24))
  Downloading mkdocstrings-0.26.1-py3-none-any.whl.metadata (8.3 kB)
Collecting netaddr==1.3.0 (from -r requirements.txt (line 25))
  Downloading netaddr-1.3.0-py3-none-any.whl.metadata (5.0 kB)
Collecting nh3==0.2.18 (from -r requirements.txt (line 26))
  Downloading nh3-0.2.18-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.7 kB)
Collecting Pillow==10.4.0 (from -r requirements.txt (line 27))
  Downloading pillow-10.4.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (9.2 kB)
Collecting psycopg==3.2.1 (from psycopg[c,pool]==3.2.1->-r requirements.txt (line 28))
  Downloading psycopg-3.2.1-py3-none-any.whl.metadata (4.2 kB)
Collecting PyYAML==6.0.2 (from -r requirements.txt (line 29))
  Downloading PyYAML-6.0.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)
Collecting requests==2.32.3 (from -r requirements.txt (line 30))
  Downloading requests-2.32.3-py3-none-any.whl.metadata (4.6 kB)
Collecting social-auth-app-django==5.4.2 (from -r requirements.txt (line 31))
  Downloading social_auth_app_django-5.4.2-py3-none-any.whl.metadata (3.2 kB)
Collecting social-auth-core==4.5.4 (from -r requirements.txt (line 32))
  Downloading social_auth_core-4.5.4-py3-none-any.whl.metadata (4.1 kB)
Collecting strawberry-graphql==0.240.2 (from -r requirements.txt (line 33))
  Downloading strawberry_graphql-0.240.2-py3-none-any.whl.metadata (7.7 kB)
Collecting strawberry-graphql-django==0.47.2 (from -r requirements.txt (line 34))
  Downloading strawberry_graphql_django-0.47.2-py3-none-any.whl.metadata (4.9 kB)
Collecting svgwrite==1.4.3 (from -r requirements.txt (line 35))
  Using cached svgwrite-1.4.3-py3-none-any.whl.metadata (8.8 kB)
Collecting tablib==3.6.1 (from -r requirements.txt (line 36))
  Downloading tablib-3.6.1-py3-none-any.whl.metadata (3.8 kB)
Collecting tzdata==2024.1 (from -r requirements.txt (line 37))
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting asgiref<4,>=3.7.0 (from Django==5.0.9->-r requirements.txt (line 1))
  Downloading asgiref-3.8.1-py3-none-any.whl.metadata (9.3 kB)
Collecting sqlparse>=0.3.1 (from Django==5.0.9->-r requirements.txt (line 1))
  Downloading sqlparse-0.5.1-py3-none-any.whl.metadata (3.9 kB)
Collecting graphene-django>=2.0.0 (from django-graphiql-debug-toolbar==0.2.0->-r requirements.txt (line 6))
  Downloading graphene_django-3.2.2-py2.py3-none-any.whl.metadata (8.3 kB)
Collecting django-js-asset (from django-mptt==0.16.0->-r requirements.txt (line 7))
  Using cached django_js_asset-2.2.0-py3-none-any.whl.metadata (3.5 kB)
Collecting six>=1.0.0 (from django-pglocks==1.0.4->-r requirements.txt (line 8))
  Using cached six-1.16.0-py2.py3-none-any.whl.metadata (1.8 kB)
Collecting prometheus-client>=0.7 (from django-prometheus==2.3.1->-r requirements.txt (line 9))
  Downloading prometheus_client-0.21.0-py3-none-any.whl.metadata (1.8 kB)
Collecting redis!=4.0.0,!=4.0.1,>=3 (from django-redis==5.4.0->-r requirements.txt (line 10))
  Downloading redis-5.0.8-py3-none-any.whl.metadata (9.2 kB)
Collecting rich>=10 (from django-rich==1.11.0->-r requirements.txt (line 11))
  Downloading rich-13.8.1-py3-none-any.whl.metadata (18 kB)
Collecting rq>=1.14 (from django-rq==2.10.2->-r requirements.txt (line 12))
  Downloading rq-1.16.2-py3-none-any.whl.metadata (5.7 kB)
Collecting uritemplate>=2.0.0 (from drf-spectacular==0.27.2->-r requirements.txt (line 17))
  Using cached uritemplate-4.1.1-py2.py3-none-any.whl.metadata (2.9 kB)
Collecting jsonschema>=2.6.0 (from drf-spectacular==0.27.2->-r requirements.txt (line 17))
  Downloading jsonschema-4.23.0-py3-none-any.whl.metadata (7.9 kB)
Collecting inflection>=0.3.1 (from drf-spectacular==0.27.2->-r requirements.txt (line 17))
  Using cached inflection-0.5.1-py2.py3-none-any.whl.metadata (1.7 kB)
Collecting sgmllib3k (from feedparser==6.0.11->-r requirements.txt (line 19))
  Using cached sgmllib3k-1.0.0.tar.gz (5.8 kB)
  Preparing metadata (setup.py) ... done
Collecting packaging (from gunicorn==23.0.0->-r requirements.txt (line 20))
  Downloading packaging-24.1-py3-none-any.whl.metadata (3.2 kB)
Collecting MarkupSafe>=2.0 (from Jinja2==3.1.4->-r requirements.txt (line 21))
  Downloading MarkupSafe-2.1.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.0 kB)
Collecting babel~=2.10 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading babel-2.16.0-py3-none-any.whl.metadata (1.5 kB)
Collecting colorama~=0.4 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Using cached colorama-0.4.6-py2.py3-none-any.whl.metadata (17 kB)
Collecting mkdocs-material-extensions~=1.3 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Using cached mkdocs_material_extensions-1.3.1-py3-none-any.whl.metadata (6.9 kB)
Collecting mkdocs~=1.6 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading mkdocs-1.6.1-py3-none-any.whl.metadata (6.0 kB)
Collecting paginate~=0.5 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading paginate-0.5.7-py2.py3-none-any.whl.metadata (11 kB)
Collecting pygments~=2.16 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading pygments-2.18.0-py3-none-any.whl.metadata (2.5 kB)
Collecting pymdown-extensions~=10.2 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading pymdown_extensions-10.10.1-py3-none-any.whl.metadata (3.0 kB)
Collecting regex>=2022.4 (from mkdocs-material==9.5.34->-r requirements.txt (line 23))
  Downloading regex-2024.9.11-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (40 kB)
Collecting click>=7.0 (from mkdocstrings==0.26.1->mkdocstrings[python-legacy]==0.26.1->-r requirements.txt (line 24))
  Using cached click-8.1.7-py3-none-any.whl.metadata (3.0 kB)
Collecting mkdocs-autorefs>=1.2 (from mkdocstrings==0.26.1->mkdocstrings[python-legacy]==0.26.1->-r requirements.txt (line 24))
  Downloading mkdocs_autorefs-1.2.0-py3-none-any.whl.metadata (8.7 kB)
Collecting platformdirs>=2.2 (from mkdocstrings==0.26.1->mkdocstrings[python-legacy]==0.26.1->-r requirements.txt (line 24))
  Downloading platformdirs-4.3.6-py3-none-any.whl.metadata (11 kB)
Collecting typing-extensions>=4.4 (from psycopg==3.2.1->psycopg[c,pool]==3.2.1->-r requirements.txt (line 28))
  Downloading typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
Collecting charset-normalizer<4,>=2 (from requests==2.32.3->-r requirements.txt (line 30))
  Downloading charset_normalizer-3.3.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (33 kB)
Collecting idna<4,>=2.5 (from requests==2.32.3->-r requirements.txt (line 30))
  Downloading idna-3.10-py3-none-any.whl.metadata (10 kB)
Collecting urllib3<3,>=1.21.1 (from requests==2.32.3->-r requirements.txt (line 30))
  Downloading urllib3-2.2.3-py3-none-any.whl.metadata (6.5 kB)
Collecting certifi>=2017.4.17 (from requests==2.32.3->-r requirements.txt (line 30))
  Downloading certifi-2024.8.30-py3-none-any.whl.metadata (2.2 kB)
Collecting oauthlib>=1.0.3 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Using cached oauthlib-3.2.2-py3-none-any.whl.metadata (7.5 kB)
Collecting requests-oauthlib>=0.6.1 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Downloading requests_oauthlib-2.0.0-py2.py3-none-any.whl.metadata (11 kB)
Collecting PyJWT>=2.7.0 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Downloading PyJWT-2.9.0-py3-none-any.whl.metadata (3.0 kB)
Collecting cryptography>=1.4 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Downloading cryptography-43.0.1-cp39-abi3-manylinux_2_28_x86_64.whl.metadata (5.4 kB)
Collecting defusedxml>=0.5.0rc1 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Using cached defusedxml-0.8.0rc2-py2.py3-none-any.whl.metadata (33 kB)
Collecting python3-openid>=3.0.10 (from social-auth-core==4.5.4->-r requirements.txt (line 32))
  Using cached python3_openid-3.2.0-py3-none-any.whl.metadata (1.6 kB)
Collecting graphql-core<3.4.0,>=3.2.0 (from strawberry-graphql==0.240.2->-r requirements.txt (line 33))
  Downloading graphql_core-3.2.4-py3-none-any.whl.metadata (10 kB)
Collecting python-dateutil<3.0.0,>=2.7.0 (from strawberry-graphql==0.240.2->-r requirements.txt (line 33))
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting mkdocstrings-python-legacy>=0.2.1 (from mkdocstrings[python-legacy]==0.26.1->-r requirements.txt (line 24))
  Downloading mkdocstrings_python_legacy-0.2.4-py3-none-any.whl.metadata (5.2 kB)
Collecting psycopg-c==3.2.1 (from psycopg[c,pool]==3.2.1->-r requirements.txt (line 28))
  Downloading psycopg_c-3.2.1.tar.gz (598 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 598.2/598.2 kB 6.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      running dist_info
      creating /tmp/pip-modern-metadata-vmnklucl/psycopg_c.egg-info
      writing /tmp/pip-modern-metadata-vmnklucl/psycopg_c.egg-info/PKG-INFO
      writing dependency_links to /tmp/pip-modern-metadata-vmnklucl/psycopg_c.egg-info/dependency_links.txt
      writing top-level names to /tmp/pip-modern-metadata-vmnklucl/psycopg_c.egg-info/top_level.txt
      writing manifest file '/tmp/pip-modern-metadata-vmnklucl/psycopg_c.egg-info/SOURCES.txt'
      couldn't run 'pg_config' --includedir: [Errno 2] No such file or directory: 'pg_config'
      error: [Errno 2] No such file or directory: 'pg_config'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
[root@netbox netbox]#
`

**Comment by jeremystretch** [comment]
It looks like you're missing the development libraries for PostgreSQL. Try `yum install libpq-devel` and then run the upgrade script again.

**Reply by bytelock-id** [reply, marked as answer]
I had to install both postgresql-devel and python3.11-devel as Python.h was also missing. 

dnf install postgresql-devel
dnf install python3.11-devel 

It did upgrade after these two packages were installed. Hopefully this will give some help to RHEL folks.
